# Analysis on the dataset of used cars from eBay Kleinanzeigen

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Observation

The dataset contains 20 columns, most of which are strings.
Some columns have null values, but none have more than ~20% null values.
The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

## Step2

Rename the column names from camelcase to snakecase.

In [5]:
corrections = {
    'dateCrawled': 'date_crawled',
    'offerType': 'offer_type',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year',
    'powerPS': 'power_ps',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'nr_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen'
}

autos.rename(columns = corrections, inplace=True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [6]:
autos.describe(include='all')
# autos['seller'].value_counts()

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

## Step3

### Investigation
- 'seller' and 'offer_type' have only 2 unique values and 1 value appears only once. 'abtest', 'gearbox', and 'unrepaired_damage' have also only 2 unique values.
- 'price' and 'odometer' columns are numeric values stored as text.

Improve the 'price' and 'odometer' columns

In [7]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)

# autos['price'].value_counts()
# autos.describe(include='all')
# autos.info()

In [8]:
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(int)

# autos['odometer'].value_counts()
# autos.info()

In [9]:
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)

# autos.info()

## Step4

Focusing on 'odometer_km' and 'price' and analyzing them.

In [10]:
autos['odometer_km'].unique().shape
autos['odometer_km'].describe()
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [11]:
autos['price'].unique().shape
autos['price'].describe()
autos['price'].value_counts().sort_index(ascending=False).head(25)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
Name: price, dtype: int64

The price above 900000 is too high to sell as a used car. These cars seem to be special cars or data bugs. Therefore, I remove these data.

In [12]:
autos = autos[autos['price'].between(0, 900000)]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49986 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        49986 non-null  object
 1   name                49986 non-null  object
 2   seller              49986 non-null  object
 3   offer_type          49986 non-null  object
 4   price               49986 non-null  int64 
 5   abtest              49986 non-null  object
 6   vehicle_type        44894 non-null  object
 7   registration_year   49986 non-null  int64 
 8   gearbox             47310 non-null  object
 9   power_ps            49986 non-null  int64 
 10  model               47233 non-null  object
 11  odometer_km         49986 non-null  int64 
 12  registration_month  49986 non-null  int64 
 13  fuel_type           45509 non-null  object
 14  brand               49986 non-null  object
 15  unrepaired_damage   40163 non-null  object
 16  ad_created          49

## Step5

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

'date_crawled' 

In [13]:
autos['date_crawled'].str[:10]
# autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()
# autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11 00:00:00    0.000020
2015-08-10 00:00:00    0.000020
2015-09-09 00:00:00    0.000020
2015-11-10 00:00:00    0.000020
2015-12-05 00:00:00    0.000020
                         ...   
2016-04-03 00:00:00    0.038931
2016-04-04 00:00:00    0.036850
2016-04-05 00:00:00    0.011843
2016-04-06 00:00:00    0.003261
2016-04-07 00:00:00    0.001280
Name: ad_created, Length: 76, dtype: float64

### Observation

- 'date_crawled' and 'last_seen' are uniformly distributed and there is no repetition.
- 'ad_created' has many data from 2016-03 to 2016-04.

In [14]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Observation

'registration_year' has outlier, which is minimum 1000 and maximum 9999.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

## Step6

In [15]:
autos['registration_year'].value_counts(normalize=True).tail(50)

1968    0.00052
1973    0.00050
1974    0.00048
1966    0.00044
1977    0.00044
1975    0.00038
1969    0.00038
1965    0.00034
1964    0.00024
1910    0.00018
1963    0.00018
1959    0.00014
1961    0.00012
1956    0.00010
5000    0.00008
9999    0.00008
1958    0.00008
1962    0.00008
1937    0.00008
2019    0.00006
1950    0.00006
1955    0.00004
9000    0.00004
1954    0.00004
1800    0.00004
1957    0.00004
1941    0.00004
1951    0.00004
1934    0.00004
4100    0.00002
4800    0.00002
1953    0.00002
1111    0.00002
1927    0.00002
6200    0.00002
4500    0.00002
1943    0.00002
5911    0.00002
1939    0.00002
1938    0.00002
2800    0.00002
8888    0.00002
1000    0.00002
1500    0.00002
1948    0.00002
1931    0.00002
1929    0.00002
1001    0.00002
9996    0.00002
1952    0.00002
Name: registration_year, dtype: float64

In [16]:
autos.loc[autos['registration_year']<1910, 'registration_year']

10556    1800
22316    1000
24511    1111
32585    1800
35238    1500
49283    1001
Name: registration_year, dtype: int64

I will extract the data between 1910 to 2016 from 'registration_year'.

In [17]:
autos = autos[autos['registration_year'].between(1910, 2016)]

In [18]:
autos['registration_year'].value_counts(normalize=True)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

## Step7

In [19]:
autos['brand'].unique()
len(autos['brand'].unique())

40

In [20]:
brand_bool = autos['brand'].value_counts(normalize=True) > 0.01
major_brands = autos['brand'].value_counts(normalize=True).index[brand_bool]

Set the 'major_brands' as the brands whose shares are more than 1% of total.

The brands are ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos']

In [29]:
mean_prices = {}
for b in major_brands:
    one_mean = autos.loc[autos['brand'] == b, 'price'].mean()
    mean_prices[b] = np.round(one_mean, 1)
mean_prices

{'volkswagen': 5231.1,
 'bmw': 8102.5,
 'opel': 2876.7,
 'mercedes_benz': 8485.2,
 'audi': 9093.7,
 'ford': 3652.1,
 'renault': 2395.4,
 'peugeot': 3039.5,
 'fiat': 2711.8,
 'seat': 4296.5,
 'skoda': 6334.9,
 'mazda': 4010.8,
 'nissan': 4664.9,
 'smart': 3542.7,
 'citroen': 3699.9,
 'toyota': 5115.3,
 'sonstige_autos': 10805.1}

In [34]:
mp_series = pd.Series(mean_prices)
print(mp_series)
df = pd.DataFrame(mp_series, columns=['mean_price'])
print(df)

volkswagen         5231.1
bmw                8102.5
opel               2876.7
mercedes_benz      8485.2
audi               9093.7
ford               3652.1
renault            2395.4
peugeot            3039.5
fiat               2711.8
seat               4296.5
skoda              6334.9
mazda              4010.8
nissan             4664.9
smart              3542.7
citroen            3699.9
toyota             5115.3
sonstige_autos    10805.1
dtype: float64
                mean_price
volkswagen          5231.1
bmw                 8102.5
opel                2876.7
mercedes_benz       8485.2
audi                9093.7
ford                3652.1
renault             2395.4
peugeot             3039.5
fiat                2711.8
seat                4296.5
skoda               6334.9
mazda               4010.8
nissan              4664.9
smart               3542.7
citroen             3699.9
toyota              5115.3
sonstige_autos     10805.1


In [41]:
mean_mile = {}
mean_price = {}
for b in major_brands:
    one_mean_price = autos.loc[autos['brand'] == b, 'price'].mean()
    one_mean_mile = autos.loc[autos['brand'] == b, 'odometer_km'].mean()
    mean_price[b] = np.round(one_mean_price, 1)
    mean_mile[b] = np.round(one_mean_mile, 1)

print(mean_mile)
print()
print(mean_price)

{'volkswagen': 128724.1, 'bmw': 132431.4, 'opel': 129223.1, 'mercedes_benz': 130856.1, 'audi': 129287.8, 'ford': 124068.7, 'renault': 128183.8, 'peugeot': 127136.8, 'fiat': 116553.9, 'seat': 121563.6, 'skoda': 110954.5, 'mazda': 124745.5, 'nissan': 118572.4, 'smart': 99595.8, 'citroen': 119580.8, 'toyota': 115709.5, 'sonstige_autos': 87466.5}

{'volkswagen': 5231.1, 'bmw': 8102.5, 'opel': 2876.7, 'mercedes_benz': 8485.2, 'audi': 9093.7, 'ford': 3652.1, 'renault': 2395.4, 'peugeot': 3039.5, 'fiat': 2711.8, 'seat': 4296.5, 'skoda': 6334.9, 'mazda': 4010.8, 'nissan': 4664.9, 'smart': 3542.7, 'citroen': 3699.9, 'toyota': 5115.3, 'sonstige_autos': 10805.1}


In [56]:
mm = pd.Series(mean_mile).sort_values(ascending=False)
print(mm)
print('\n')
mp = pd.Series(mean_price).sort_values(ascending=False)
print(mp)

bmw               132431.4
mercedes_benz     130856.1
audi              129287.8
opel              129223.1
volkswagen        128724.1
renault           128183.8
peugeot           127136.8
mazda             124745.5
ford              124068.7
seat              121563.6
citroen           119580.8
nissan            118572.4
fiat              116553.9
toyota            115709.5
skoda             110954.5
smart              99595.8
sonstige_autos     87466.5
dtype: float64


sonstige_autos    10805.1
audi               9093.7
mercedes_benz      8485.2
bmw                8102.5
skoda              6334.9
volkswagen         5231.1
toyota             5115.3
nissan             4664.9
seat               4296.5
mazda              4010.8
citroen            3699.9
ford               3652.1
smart              3542.7
peugeot            3039.5
opel               2876.7
fiat               2711.8
renault            2395.4
dtype: float64


In [57]:
df = pd.DataFrame(mm, columns=['mean_mile'])
df['mean_price'] = mp
df

mean_mile  mean_price
bmw              132431.4      8102.5
mercedes_benz    130856.1      8485.2
audi             129287.8      9093.7
opel             129223.1      2876.7
volkswagen       128724.1      5231.1
renault          128183.8      2395.4
peugeot          127136.8      3039.5
mazda            124745.5      4010.8
ford             124068.7      3652.1
seat             121563.6      4296.5
citroen          119580.8      3699.9
nissan           118572.4      4664.9
fiat             116553.9      2711.8
toyota           115709.5      5115.3
skoda            110954.5      6334.9
smart             99595.8      3542.7
sonstige_autos    87466.5     10805.1